### PR TITLE
Attention sink support

### PR DESCRIPTION
**Pull request by kareemshaik80**
- Support for single sink logit in flash attention Decode
 - Add Sink to Softmax
 - Cmd line flag added to enable attention sink

### CODE REVIEW

File: applications/flash_attention_v2/kernel/xe_flash_attn_decode.hpp [hunk 1]
```diff
@@ -340,7 +342,14 @@ class FMHADecode {
       CollectiveMainloop collective_mma;
 
       ElementAccumulator max_reg = ElementAccumulator{-INFINITY};
+      ElementAccumulator sink_token = ElementAccumulator{ 0 };
       auto sum_reg = ElementAccumulator{0};
+      if constexpr (HasSink) {
+        if (syclcompat::get_nd_item<3>().get_local_linear_id() == 0) {
```
Comment (yuankuns) [inline applications/flash_attention_v2/kernel/xe_flash_attn_decode.hpp:348]: Why specific linear id need for calculation?

File: applications/flash_attention_v2/kernel/xe_flash_attn_decode.hpp [hunk 2]
```diff
@@ -340,7 +342,14 @@ class FMHADecode {
...
+          max_reg = static_cast<ElementAccumulator>(mainloop_params.ptr_Sink[num_heads_coord]);
```
Comment (yuankuns) [inline applications/flash_attention_v2/kernel/xe_flash_attn_decode.hpp:349]: ptr operation is not recommended

**Review by yuankuns — CHANGES_REQUESTED** [review]
Also need paper/code reference to ensure this PR is what intended to do

File: applications/flash_attention_v2/kernel/xe_flash_attn_decode.hpp (hunk 1 above)
Comment (kareemshaik80) [reply applications/flash_attention_v2/kernel/xe_flash_attn_decode.hpp:348]: In online-softMax, we will initialize Max value with sink token and this should happen once per Thread Block. Out of all threads in a work group, will use the first thread ( liner ID is zero ) for the initialization.

File: applications/flash_attention_v2/kernel/xe_flash_attn_decode.hpp (hunk 2 above)
Comment (kareemshaik80) [reply applications/flash_attention_v2/kernel/xe_flash_attn_decode.hpp:349]: This is basically pointing to sink tensor not a pointer. What we are doing here is that we are using per head sink token across thread Block.

**Review by kareemshaik80 — COMMENTED** [review]
Move test under unit tests.

File: applications/flash_attention_v2/collective/xe_flash_attn_decode_softmax_epilogue.hpp [hunk 3]
```diff
@@ -162,6 +163,13 @@ class FlashDecodeSoftmaxEpilogue<CausalMask_, epilogue::IntelXeXMX16, Element_>
 
     reduce_max<Num_SGs,FragsNS>(frag_s, shmem_tensor_max, max_val);
 
+    if constexpr (HasSink) {
+      if (compat::get_nd_item<3>().get_local_linear_id() == 0) {
+        Element max_scale{max_val * params.scale};
+        sum += sycl::native::exp2((sink_token- max_scale));
```
[the remark this answers is not in the document]
Comment (kareemshaik80) [reply applications/flash_attention_v2/collective/xe_flash_attn_decode_softmax_epilogue.hpp]: Yes. we do max and sum only in thread 0.

**Comment by sunjiweiswift** [comment]
The calculation here is incorrect.
1. We use exp2, so we need to add a sink x (constexpr double kLog2e = 1.4426950408889634074);

2. In oneline softmax, it's more appropriate to process the sink in stage 2, as is done in triton (https://github.com/openai/gpt-oss/blob/0a9ec7f69d8aa71841c5cefcd84a512344b9f1be/gpt_oss/triton/attention.py#L94C4-L100C46). Introducing the sink in stage 1 is incorrect; it will change and gemm results for V

I've completed this functionality in cutlass,   https://github.com/intel/sycl-tla/pull/498/commits/f709e32e45dcbfd8897fcfe5212790bf07a55347


It is recommended to use a more stringent ut to check the results
